lib/Subscriber/Core/MappingSubscriber.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+namespace Sulu\Component\DocumentManager\Subscriber\Core;
+
+use Sulu\Component\DocumentManager\PropertyEncoder;
+use Sulu\Component\DocumentManager\Event\HydrateEvent;
+use Sulu\Component\DocumentManager\Event\PersistEvent;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Sulu\Component\DocumentManager\MetadataFactoryInterface;
+use Sulu\Component\DocumentManager\Events;
+
+class MappingSubscriber implements EventSubscriberInterface
+{
+    private $factory;
+    private $encoder;
+
+    public function __construct(
+        MetadataFactoryInterface $factory,
+        PropertyEncoder $encoder
+    )
+    {
+        $this->factory = $factory;
+        $this->encoder = $encoder;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    public static function getSubscribedEvents()
+    {
+        return array(
+            Events::HYDRATE => array('handleHydrate', 200),
+            Events::PERSIST => array('handlePersist', 200),
+        );
+    }
+
+    public function handlePersist(PersistEvent $event)
+    {
+        $metadata = $this->factory->getMetadataForClass(get_class($event->getDocument()));
+        $locale = $event->getLocale();
+        $node = $event->getNode();
+
+        foreach ($metadata->getFieldMappings() as $fieldName => $fieldMapping) {
+            $encoding = $fieldMapping['encoding'];
+            $name = $fieldMapping['property'];
+            $phpcrName = $this->encoder->encode($encoding, $name, $locale);
+            $value = $event->getAccessor()->get($fieldName);
+
+            $node->setProperty($phpcrName, $value);
+        }
+    }
+
+    public function handleHydrate(HydrateEvent $event)
+    {
+        $metadata = $this->factory->getMetadataForClass(get_class($event->getDocument()));
+        $locale = $event->getLocale();
+        $node = $event->getNode();
+
+        foreach ($metadata->getFieldMappings() as $fieldName => $fieldMapping) {
+            $encoding = $fieldMapping['encoding'];
+            $name = $fieldMapping['property'];
+            $phpcrName = $this->encoder->encode($encoding, $name, $locale);
+            $value = $node->getPropertyValueWithDefault($phpcrName, null);
+            $event->getAccessor()->set($fieldName, $value);
```

Comment:
Here we should try and add a `ChangeTracker` here to only write values that have changed. And then we should benchmark it to see what performance impact it has.